Split out to modules, and other things

mgwalker opened the pull request:
I should have made this PR sooner so it'd be smaller, but here we are...
### Modules

Broke out the major functional areas into modules, and then broke some of those down as well.
- `bot-joinChannel` - just handles messaging the channel when joined
- `bot-createStandup` - handles standup creation messages
- `bot-replyToBadStandup` - puts together a useful message letting the user know their standup wasn't created and why
- `bot-getUserStandupInfo` - handles standup user messages
- `bot-runReports` - check for which channels need their reports run based on the current time, and run them
- `helper-time` - wraps up all the time format strings in one place, and does some basic validation; also houses the schedule time regex
### Lint

Added a dependency on jshint, and an npm script to run it:  `npm run lint`.  Uses the .jshintrc file created by @stvnrlly (from airbnb).
### Open Questions
- Does the module naming convention make sense?  Currently modules that have knowledge about the bot are prefixed with `bot-` and helper libraries prefixed with `helper-`.  It's pretty wordy and I'm not convinced it's helpful.
- In the `app.js` kickoff, the bot does nothing until the `identifyBot()` method returns.  This was my change, to get the bot and its name in scope and get rid of some global variables.  The downside is the extra level of nesting in the code.  Should that be reverted?

mgwalker committed:
Break out major functional pieces to modules

mgwalker committed:
Make it pass lint

mgwalker committed:
• Add another module
• Rename modules to prefix them with functional units (we can discuss this before merging)
• Moved actual bot setup to inside the identifyBot callback, so everything is available locally; this also mitigates a situation where the identifyBot function fails but all the handlers are hooked up anywhere.  This is a pretty minor situation (the join-channel message is incomplete) so I'm not set on this at all.

mgwalker committed:
Carry that rename throughout, howsabout?

mgwalker committed:
Add a time helper module, to collect all the date/time format strings in one place

mgwalker committed:
Detect when the time, frequency, whatever is empty and report that rather than crashing.

mgwalker committed:
Refactored a bit to chain promises rather than nest them, for whatever that's worth

mgwalker committed:
Lowercase the test value, since the expected values are also lowercase

mgwalker committed:
Lowercase before taking the first character, in case of an empty string

mgwalker committed:
Link to the channel

mgwalker committed:
Verify the time is valid - entering "2500" is invalid, for example

mgwalker committed:
Moar modules!

mgwalker committed:
Commented-out variables can be tossed

mgwalker committed:
Name was off

stvnrlly commented:
Nice! I'll test this out locally when I have a moment.

---

My thoughts on structure are largely me trying to imagine myself as an outsider trying to figure out the code. With that in mind, here are my feelings:

- Grouping related functions is great. A clearly-defined helper is also a big plus. For further grouping, I kind of like the approach used [here](https://github.com/18F/analytics-reporter/blob/master/analytics.js) where the related functions are all parts of a set.

- I'd rather avoid the nesting, for no other reason than it more clearly follows the examples given by the [botkit](https://github.com/howdyai/botkit) library we're using.

mgwalker commented:
On your first point, I don't know why it didn't occur to me in the first place to follow what you did with the `models` directory - throw an index in there and let it handle importing the rest.  It simplifies things from the outside.  It'd be great for cleaning up some of the code I dislike the most - the `require('whatever')(args)` stuff.  I didn't like it when I wrote it and then forgot about it so I never made it better.

I agree on the second point as well.  I'm curious now, though, if the bot should wait until the `identifyBot()` method returns - if it gets an interaction between the time it connects and when it gets its identity, it could make for a weird user experience.  But that's a short window and a minor issue, so `¯\_(ツ)_/¯`.

mgwalker committed:
Group lib functions into component parts

mgwalker committed:
Rename runReports to getReportRunner to more accurately reflect what's going on

mgwalker commented:
Added `lib/bot` and `lib/helpers` directories, and moved the corresponding modules into them.  Added an `index.js` to each, so you can now `require('./lib/bot')` and get access to all of the bot lib methods.  Updated the rest of the code accordingly.
